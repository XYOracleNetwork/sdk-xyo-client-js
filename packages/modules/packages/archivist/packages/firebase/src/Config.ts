import { ArchivistConfig } from '@xyo-network/archivist-model'

import { FirebaseArchivistSchema } from './Schema.ts'

export type FirebaseArchivistConfigSchema = `${FirebaseArchivistSchema}.config`
export const FirebaseArchivistConfigSchema: FirebaseArchivistConfigSchema = `${FirebaseArchivistSchema}.config`

export type FirebaseArchivistConfig = ArchivistConfig<{
  collection?: string
  dbId?: string
  schema: FirebaseArchivistConfigSchema
}>
