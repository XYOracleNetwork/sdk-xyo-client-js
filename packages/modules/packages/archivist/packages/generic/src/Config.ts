import type { EmptyObject, WithAdditional } from '@xylabs/object'
import type { ArchivistConfig } from '@xyo-network/archivist-model'
import type { Payload, Schema } from '@xyo-network/payload-model'

export const GenericArchivistConfigSchema = 'network.xyo.archivist.generic.config' as const
export type GenericArchivistConfigSchema = typeof GenericArchivistConfigSchema

export type GenericArchivistConfig<TConfig extends Payload | EmptyObject | void = void, TSchema extends Schema | void = void> = ArchivistConfig<
  WithAdditional<
    {
      max?: number
    },
    TConfig
  >,
  TSchema extends Schema ? TSchema : GenericArchivistConfigSchema | ArchivistConfig['schema']
>
