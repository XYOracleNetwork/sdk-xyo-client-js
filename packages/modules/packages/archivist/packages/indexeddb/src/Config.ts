import { ArchivistConfig } from '@xyo-network/archivist-model'

import { IndexedDbArchivistSchema } from './Schema'

export type IndexedDbArchivistConfigSchema = `${IndexedDbArchivistSchema}.config`
export const IndexedDbArchivistConfigSchema: IndexedDbArchivistConfigSchema = `${IndexedDbArchivistSchema}.config`

export type IndexedDbArchivistConfig = ArchivistConfig<{
  /**
   * The database name
   */
  dbName?: string
  /**
   * The version of the DB, defaults to 1
   */
  dbVersion?: number
  schema: IndexedDbArchivistConfigSchema
  /**
   * The name of the object store
   */
  storeName?: string
}>
