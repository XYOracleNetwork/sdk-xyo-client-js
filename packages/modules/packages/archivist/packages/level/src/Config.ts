import type { ArchivistConfig } from '@xyo-network/archivist-model'

import { LevelDbArchivistSchema } from './Schema.ts'

export type LevelDbArchivistConfigSchema = `${LevelDbArchivistSchema}.config`
export const LevelDbArchivistConfigSchema: LevelDbArchivistConfigSchema = `${LevelDbArchivistSchema}.config`

export type LevelDbArchivistConfig<TStoreName extends string = string> = ArchivistConfig<{
  /**
   * The database name
   */
  dbName?: string
  location?: string
  schema: LevelDbArchivistConfigSchema
  /**
   * The name of the object store
   */
  storeName?: TStoreName
}>
