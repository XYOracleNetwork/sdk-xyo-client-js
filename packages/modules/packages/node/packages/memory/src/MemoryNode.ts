import { assertEx } from '@xylabs/assert'
import { AnyConfigSchema, CompositeModuleResolver, EventListener, Module, ModuleInstance, ModuleResolver } from '@xyo-network/module'
import { AbstractNode } from '@xyo-network/node-abstract'
import { isNodeModule, NodeConfig, NodeConfigSchema, NodeInstance, NodeModuleEventData, NodeModuleParams } from '@xyo-network/node-model'
import compact from 'lodash/compact'

export type MemoryNodeParams = NodeModuleParams<AnyConfigSchema<NodeConfig>>

export class MemoryNode<TParams extends MemoryNodeParams = MemoryNodeParams, TEventData extends NodeModuleEventData = NodeModuleEventData>
  extends AbstractNode<TParams, TEventData>
  implements NodeInstance
{
  static override configSchemas = [NodeConfigSchema]

  private registeredModuleMap: Record<string, ModuleInstance> = {}

  override async attach(nameOrAddress: string, external?: boolean) {
    await this.started('throw')
    return (await this.attachUsingAddress(nameOrAddress, external)) ?? (await this.attachUsingName(nameOrAddress, external))
  }

  override async detach(nameOrAddress: string) {
    await this.started('throw')
    return (await this.detachUsingAddress(nameOrAddress)) ?? (await this.detachUsingName(nameOrAddress))
  }

  override async register(module: ModuleInstance) {
    await this.started('throw')
    assertEx(!this.registeredModuleMap[module.address], `Module already registered at that address[${module.address}][${module.config.schema}]`)
    this.registeredModuleMap[module.address] = module
    const args = { module, name: module.config.name }
    await this.emit('moduleRegistered', args)
  }

  override registered() {
    return Object.keys(this.registeredModuleMap).map((key) => {
      return key
    })
  }

  override registeredModules() {
    return Object.values(this.registeredModuleMap).map((value) => {
      return value
    })
  }

  override async unregister(module: ModuleInstance) {
    await this.started('throw')
    await this.detach(module.address)
    delete this.registeredModuleMap[module.address]
    const args = { module, name: module.config.name }
    await this.emit('moduleUnregistered', args)
    return this
  }

  protected override startHandler() {
    return super.startHandler()
  }

  private async attachUsingAddress(address: string, external?: boolean) {
    const existingModule = (await this.resolve({ address: [address] })).pop()
    assertEx(!existingModule, `Module [${existingModule?.config.name ?? existingModule?.address}] already attached at address [${address}]`)
    const module = this.registeredModuleMap[address]

    if (!module) {
      return
    }

    const notifiedAddresses: string[] = []

    const getModulesToNotifyAbout = async (node: ModuleInstance) => {
      //send attach events for all existing attached modules
      const childModules = await node.resolve(undefined, { direction: 'down' })
      return compact(
        childModules.map((child) => {
          //don't report self
          if (node.address === child.address) {
            return
          }

          //prevent loop
          if (notifiedAddresses.includes(child.address)) {
            return
          }

          notifiedAddresses.push(child.address)

          return child
        }),
      )
    }

    const notificationList = await getModulesToNotifyAbout(module)

    //give it private access
    module.upResolver.addResolver?.(this.privateResolver)

    //give it public access
    module.upResolver.addResolver?.(this.downResolver as CompositeModuleResolver)

    //give it outside access
    module.upResolver.addResolver?.(this.upResolver)

    if (external) {
      //expose it externally
      this.downResolver.addResolver(module.downResolver as ModuleResolver)
    } else {
      this.privateResolver.addResolver(module.downResolver as ModuleResolver)
    }

    const args = { module, name: module.config.name }
    await this.emit('moduleAttached', args)

    if (isNodeModule(module)) {
      if (external) {
        const attachedListener: EventListener<TEventData['moduleAttached']> = async (args: TEventData['moduleAttached']) =>
          await this.emit('moduleAttached', args)

        const detachedListener: EventListener<TEventData['moduleDetached']> = async (args: TEventData['moduleDetached']) =>
          await this.emit('moduleDetached', args)

        module.on('moduleAttached', attachedListener)
        module.on('moduleDetached', detachedListener)
      }
    }

    const notifyOfExistingModules = async (childModules: Module[]) => {
      await Promise.all(
        childModules.map(async (child) => {
          const args = { module: child, name: child.config.name }
          await this.emit('moduleAttached', args)
        }),
      )
    }

    await notifyOfExistingModules(notificationList)

    return address
  }

  private async attachUsingName(name: string, external?: boolean) {
    const address = this.moduleAddressFromName(name)
    if (address) {
      return await this.attachUsingAddress(address, external)
    }
  }

  private async detachUsingAddress(address: string) {
    const module = this.registeredModuleMap[address]

    if (!module) {
      return
    }

    //remove inside access
    module.upResolver?.removeResolver?.(this.privateResolver)

    //remove outside access
    module.upResolver?.removeResolver?.(this.upResolver)

    //remove external exposure
    this.downResolver.removeResolver(module.downResolver as ModuleResolver)

    const args = { module, name: module.config.name }
    await this.emit('moduleDetached', args)

    //notify of all sub node children detach
    const notifiedAddresses: string[] = []
    if (isNodeModule(module)) {
      const notifyOfExistingModules = async (node: ModuleInstance) => {
        //send attach events for all existing attached modules
        const childModules = await node.resolve(undefined, { direction: 'down' })
        await Promise.all(
          childModules.map(async (child) => {
            //don't report self
            if (node.address === child.address) {
              return
            }

            //prevent loop
            if (notifiedAddresses.includes(child.address)) {
              return
            }
            notifiedAddresses.push(child.address)
            await this.emit('moduleDetached', { module: child })
          }),
        )
      }
      await notifyOfExistingModules(module)
    }
    return address
  }

  private async detachUsingName(name: string) {
    const address = this.moduleAddressFromName(name)
    if (address) {
      return await this.detachUsingAddress(address)
    }
    return
  }

  private moduleAddressFromName(name: string) {
    const address = Object.values(this.registeredModuleMap).find((value) => {
      return value.config.name === name
    }, undefined)?.address
    return address
  }
}
