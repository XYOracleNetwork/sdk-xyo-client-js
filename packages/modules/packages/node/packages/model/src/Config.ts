import { EmptyObject, WithAdditional } from '@xylabs/object'
import { ModuleConfig } from '@xyo-network/module-model'
import { Payload } from '@xyo-network/payload-model'

export const NodeConfigSchema = 'network.xyo.node.config'
export type NodeConfigSchema = typeof NodeConfigSchema

export type NodeConfig<TConfig extends EmptyObject | Payload | void = void, TSchema extends string | void = void> = ModuleConfig<
  WithAdditional<
    {
      archivist?: string
    },
    TConfig
  >,
  TSchema extends void ?
    TConfig extends Payload ?
      TConfig['schema']
    : NodeConfigSchema
  : TSchema
>
