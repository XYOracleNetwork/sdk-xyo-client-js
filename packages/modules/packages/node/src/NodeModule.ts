import { Module, ModuleFilter } from '@xyo-network/module'
import { Promisable, PromisableArray } from '@xyo-network/promise'

import { Node } from './Node'

export interface NodeModule<TModule extends Module = Module> extends Node, Module {
  attachedModules(): PromisableArray<TModule>
  register(module: TModule): Promisable<void>
  registeredModules(): PromisableArray<TModule>
  resolve(filter?: ModuleFilter): PromisableArray<TModule>
  unregister(module: TModule): Promisable<void>
}
