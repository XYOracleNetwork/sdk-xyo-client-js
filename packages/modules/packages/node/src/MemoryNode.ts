import { assertEx } from '@xylabs/assert'
import { XyoArchivistGetQuerySchema } from '@xyo-network/archivist'
import { XyoModule } from '@xyo-network/module'

import { NodeConfig } from './Config'
import { XyoNodeAttachQuerySchema, XyoNodeDetachQuerySchema } from './Queries'
import { XyoNode } from './XyoNode'

export class MemoryNode<TConfig extends NodeConfig = NodeConfig, TModule extends XyoModule = XyoModule> extends XyoNode<TConfig, TModule> {
  private registeredModuleMap = new Map<string, TModule>()
  private attachedModuleMap = new Map<string, TModule>()

  public override queries(): string[] {
    return [XyoNodeAttachQuerySchema, XyoNodeDetachQuerySchema, ...super.queries()]
  }

  override attached() {
    return Array.from(this.attachedModuleMap.keys()).map((key) => {
      return key
    })
  }

  override registered() {
    return Array.from(this.registeredModuleMap.keys()).map((key) => {
      return key
    })
  }

  override attachedModules() {
    return Array.from(this.attachedModuleMap.values()).map((value) => {
      return value
    })
  }

  override registeredModules() {
    return Array.from(this.registeredModuleMap.values()).map((value) => {
      return value
    })
  }

  override resolve(address: string) {
    console.log(`Resolving in MemoryNode: ${address}`)
    if (address === 'archivist') {
      console.log(`attachedModules: ${JSON.stringify(this.attachedModules(), null, 2)}`)
      return this.attachedModules().find((module) => module.queryable(XyoArchivistGetQuerySchema)) ?? null
    }
    return this.attachedModuleMap?.get(address) ?? null
  }

  override register(module: TModule) {
    this.registeredModuleMap.set(module.address, module)
  }

  override attach(address: string) {
    const module = assertEx(this.registeredModuleMap.get(address), 'No module found at that address')
    this.attachedModuleMap.set(address, module)
  }

  override detach(address: string) {
    this.attachedModuleMap.delete(address)
  }
}
