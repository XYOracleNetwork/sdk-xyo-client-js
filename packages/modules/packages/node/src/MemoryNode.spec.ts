/* eslint-disable max-statements */
import { XyoArchivistWrapper, XyoMemoryArchivist, XyoMemoryArchivistConfigSchema } from '@xyo-network/archivist'
import {
  DivinerModule,
  XyoArchivistPayloadDiviner,
  XyoArchivistPayloadDivinerConfigSchema,
  XyoDivinerWrapper,
  XyoHuriPayload,
  XyoHuriSchema,
} from '@xyo-network/diviner'
import { ModuleDescription, XyoModule, XyoModuleResolver } from '@xyo-network/module'
import { PayloadWrapper, XyoAccount, XyoPayload, XyoPayloadBuilder, XyoPayloadSchema } from '@xyo-network/sdk'

import { NodeConfigSchema } from './Config'
import { MemoryNode } from './MemoryNode'
import { NodeModule } from './NodeModule'

describe('MemoryNode', () => {
  describe('create', () => {
    it('Creates MemoryNode', async () => {
      const XyoMemoryArchivist = (await import('@xyo-network/archivist')).XyoMemoryArchivist
      const node: NodeModule = await MemoryNode.create()
      const archivist = await XyoMemoryArchivist.create()
      const diviner: XyoModule = await XyoArchivistPayloadDiviner.create({
        config: { archivist: archivist.address, schema: XyoArchivistPayloadDivinerConfigSchema },
        resolver: new XyoModuleResolver().add(archivist),
      })
      await node.register(archivist)
      node.attach(archivist.address)
      await node.register(diviner)
      node.attach(diviner.address)
      expect((await node.registered()).length).toBe(2)
      expect((await node.attached()).length).toBe(2)
      const foundArchivist = (await node.resolve({ address: [archivist.address] })).shift()
      expect(foundArchivist).toBeDefined()
      expect(foundArchivist?.address).toBe(archivist.address)
      const testPayload = new XyoPayloadBuilder<XyoPayload<{ schema: XyoPayloadSchema; test: boolean }>>({ schema: XyoPayloadSchema })
        .fields({ test: true })
        .build()

      const foundArchivistWrapper = foundArchivist ? new XyoArchivistWrapper(foundArchivist) : undefined
      await foundArchivistWrapper?.insert([testPayload])

      /*const subscribeQuery: XyoModuleSubscribeQuery = { payloads: [testPayload], schema: XyoModuleSubscribeQuerySchema }
  await foundArchivist?.query(subscribeQuery)*/

      const payloads = await foundArchivistWrapper?.all()
      expect(payloads?.length).toBe(1)

      if (payloads && payloads[0]) {
        const huri = new PayloadWrapper(payloads[0]).hash
        const huriPayload: XyoHuriPayload = { huri: [huri], schema: XyoHuriSchema }
        const module = (await node.resolve({ address: [diviner.address] })).shift() as DivinerModule
        const foundDiviner = module ? new XyoDivinerWrapper(module) : null
        expect(foundDiviner).toBeDefined()
        if (foundDiviner) {
          const foundDivinerWrapper = new XyoDivinerWrapper(foundDiviner)
          const payloads = await foundDivinerWrapper.divine([huriPayload])
          // console.log(`payloads: ${JSON.stringify(payloads, null, 2)}`)
          expect(payloads?.length).toBe(1)
          expect(payloads[0]).toBeDefined()
          if (payloads?.length === 1 && payloads[0]) {
            expect(new PayloadWrapper(payloads[0]).hash).toBe(huri)
          }
        }
      }
    })
  })
  describe('description', () => {
    const testNodeAccount = new XyoAccount({ phrase: 'testPhrase1' })
    let node: MemoryNode
    const validateModuleDescription = (description: ModuleDescription) => {
      expect(description).toBeObject()
      expect(description.address).toBeString()
      expect(description.queries).toBeArray()
      description.queries.map((query) => {
        expect(query).toBeString()
      })
    }
    describe('node without children', () => {
      beforeAll(async () => {
        node = await MemoryNode.create({ account: testNodeAccount, config: { schema: NodeConfigSchema } })
      })
      it('describes node alone', async () => {
        const description = await node.description()
        validateModuleDescription(description)
        expect(description.children).toBeArrayOfSize(0)
      })
      it('serializes to JSON consistently', async () => {
        const description = await node.description()
        expect(JSON.stringify(description)).toMatchSnapshot()
      })
    })
    describe('node with children', () => {
      const testArchivistAccount1 = new XyoAccount({ phrase: 'testPhrase2' })
      const testArchivistAccount2 = new XyoAccount({ phrase: 'testPhrase3' })
      let modules: XyoModule[]
      beforeAll(async () => {
        node = await MemoryNode.create({ account: testNodeAccount, config: { schema: NodeConfigSchema } })
        const config = { schema: XyoMemoryArchivistConfigSchema }
        modules = await Promise.all([
          await XyoMemoryArchivist.create({ account: testArchivistAccount1, config }),
          await XyoMemoryArchivist.create({ account: testArchivistAccount2, config }),
        ])
        modules.map((archivist) => {
          node.register(archivist)
          node.attach(archivist.address)
        })
      })
      it('describes node and children', async () => {
        const description = await node.description()
        validateModuleDescription(description)
        expect(description.children).toBeArrayOfSize(modules.length)
        description.children?.map(validateModuleDescription)
      })
      it('serializes to JSON consistently', async () => {
        const description = await node.description()
        expect(JSON.stringify(description)).toMatchSnapshot()
      })
    })
  })
})
