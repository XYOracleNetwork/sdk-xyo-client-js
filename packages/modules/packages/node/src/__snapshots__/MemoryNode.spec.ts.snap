// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoryNode description node with children serializes to JSON consistently 1`] = `
"{
  "address": "3a15a6c96665931b76c1d2a587ea1132dbfdc266",
  "queries": [
    "network.xyo.query.node.attach",
    "network.xyo.query.node.detach",
    "network.xyo.query.module.discover",
    "network.xyo.query.module.subscribe"
  ],
  "children": [
    {
      "address": "b3bbeb946d88cd5fe70146a6bb276c7b384f6582",
      "queries": [
        "network.xyo.query.archivist.all",
        "network.xyo.query.archivist.delete",
        "network.xyo.query.archivist.clear",
        "network.xyo.query.archivist.find",
        "network.xyo.query.archivist.commit",
        "network.xyo.query.archivist.get",
        "network.xyo.query.archivist.insert",
        "network.xyo.query.module.discover",
        "network.xyo.query.module.subscribe"
      ]
    },
    {
      "address": "ca2e94ac3ae1084072b3a4407881bf940dc2019c",
      "queries": [
        "network.xyo.query.archivist.all",
        "network.xyo.query.archivist.delete",
        "network.xyo.query.archivist.clear",
        "network.xyo.query.archivist.find",
        "network.xyo.query.archivist.commit",
        "network.xyo.query.archivist.get",
        "network.xyo.query.archivist.insert",
        "network.xyo.query.module.discover",
        "network.xyo.query.module.subscribe"
      ]
    }
  ]
}"
`;

exports[`MemoryNode description node with nested node and children serializes to JSON consistently 1`] = `
"{
  "address": "3a15a6c96665931b76c1d2a587ea1132dbfdc266",
  "queries": [
    "network.xyo.query.node.attach",
    "network.xyo.query.node.detach",
    "network.xyo.query.module.discover",
    "network.xyo.query.module.subscribe"
  ],
  "children": [
    {
      "address": "f59810a9da8a99e464355a411afe06f8c2d1c4b8",
      "queries": [
        "network.xyo.query.archivist.all",
        "network.xyo.query.archivist.delete",
        "network.xyo.query.archivist.clear",
        "network.xyo.query.archivist.find",
        "network.xyo.query.archivist.commit",
        "network.xyo.query.archivist.get",
        "network.xyo.query.archivist.insert",
        "network.xyo.query.module.discover",
        "network.xyo.query.module.subscribe"
      ]
    },
    {
      "address": "b3bbeb946d88cd5fe70146a6bb276c7b384f6582",
      "queries": [
        "network.xyo.query.node.attach",
        "network.xyo.query.node.detach",
        "network.xyo.query.module.discover",
        "network.xyo.query.module.subscribe"
      ],
      "children": [
        {
          "address": "ca2e94ac3ae1084072b3a4407881bf940dc2019c",
          "queries": [
            "network.xyo.query.archivist.all",
            "network.xyo.query.archivist.delete",
            "network.xyo.query.archivist.clear",
            "network.xyo.query.archivist.find",
            "network.xyo.query.archivist.commit",
            "network.xyo.query.archivist.get",
            "network.xyo.query.archivist.insert",
            "network.xyo.query.module.discover",
            "network.xyo.query.module.subscribe"
          ]
        }
      ]
    }
  ]
}"
`;

exports[`MemoryNode description node without children serializes to JSON consistently 1`] = `
"{
  "address": "3a15a6c96665931b76c1d2a587ea1132dbfdc266",
  "queries": [
    "network.xyo.query.node.attach",
    "network.xyo.query.node.detach",
    "network.xyo.query.module.discover",
    "network.xyo.query.module.subscribe"
  ],
  "children": []
}"
`;
